README.md
# nanomap [![stability][0]][1]
[![npm version][2]][3] [![build status][4]][5]
[![downloads][8]][9] [![js-standard-style][10]][11]

Functionally map data into stateful [nanocomponents][nc].

![](screenshot.png)

## Usage

```js
var Nanomap = require('nanocomponent/map')
var YoutubeComponent = require('youtube-component')
var TwitterComponent = require('twitter-component')
var OEmbedComponent = require('oembed-component')
var simpleMapper = new Nanomap(opts, TwitterComponent)
// OR
var complexMapper = new Nanomap(opts, {
  'video': YoutubeComponent,
  'tweet': TwitterComponent,
  ...,
  default: OEmbedComponent
})

[{
  id: 'foo123',
  opts: { color: 'blue' },
  arguments: {an: 'arg'} // Non-array types passed in as the first argument
}].map(simpleMapper) // Array of rendered DOM nodes from homogeneous components

[{
  id: 'foo123',
  type: 'tweet',
  arguments: ['tweet-url'] // component.render.apply(component, arguments)
}].map(complexMapper) // Array of rendered DOM nodes from a heterogeneous set of components
```

## Installation
```sh
$ npm install nanomap
```
## API
### `Nanomap = require('nanomap`)
Import `Nanomap` component class.

### `mapper = new Nanomap([opts], Component)`
### `mapper = Nanomap([opts], { type: Component, [default: Component]})`
Create a new mapper instance that will render data into component instances.

`opts` include:

```js
{
  gc: true // clean up unused instances when mapped over
}
```

See examples for more details.

## License
[MIT](https://tldrlegal.com/license/mit-license)

[0]: https://img.shields.io/badge/stability-experimental-orange.svg?style=flat-square
[1]: https://nodejs.org/api/documentation.html#documentation_stability_index
[2]: https://img.shields.io/npm/v/nanomap.svg?style=flat-square
[3]: https://npmjs.org/package/nanomap
[4]: https://img.shields.io/travis/bcomnes/nanomap/master.svg?style=flat-square
[5]: https://travis-ci.org/bcomnes/nanomap
[8]: http://img.shields.io/npm/dm/nanomap.svg?style=flat-square
[9]: https://npmjs.org/package/nanomap
[10]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square
[11]: https://github.com/feross/standard
[bel]: https://github.com/shama/bel
[yoyoify]: https://github.com/shama/yo-yoify
[md]: https://github.com/patrick-steele-idem/morphdom
[210]: https://github.com/patrick-steele-idem/morphdom/pull/81
[nm]: https://github.com/yoshuawuyts/nanomorph
[ce]: https://github.com/yoshuawuyts/cache-element
[class]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes
[isSameNode]: https://github.com/choojs/nanomorph#caching-dom-elements
[onload]: https://github.com/shama/on-load
[choo]: https://github.com/choojs/choo
[nca]: https://github.com/choojs/nanocomponent-adapters
[nc]: https://github.com/choojs/nanocomponent


CHANGELOG.md
# nanomap Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 1.1.1 - 2017-09-11
* Update dep floor

## 1.1.0 - 2017-08-24
* **Added**: Use [`nanoassert`](https://github.com/emilbayes/nanoassert) in the browser

## 1.0.0 - 2017-08-20
* Initial release.  Beta API basically.


test.js
var test = require('tape')
var YoutubeComponent = require('youtube-component')
var TwitterComponent = require('twitter-component')
var Nanomap = require('./')

function makeID () {
  return 'testid-' + Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1)
}

function createTestElement () {
  var testRoot = document.createElement('div')
  testRoot.id = makeID()
  document.body.appendChild(testRoot)
  return testRoot
}

function renderAndMount (testEl, data, mapper) {
  var els = data.map(mapper)
  els.forEach(function (el) {
    testEl.appendChild(el)
  })
}

test('simple mapper', function (t) {
  var testRoot = createTestElement()
  var simpleMapper = new Nanomap(YoutubeComponent)
  var videos = [
    'https://www.youtube.com/watch?v=b8HO6hba9ZE',
    'https://vimeo.com/229754542',
    'https://www.youtube.com/watch?v=bYpKrA233vY'
  ].map(function (url) {
    return { id: url, arguments: url }
  })

  t.doesNotThrow(renderAndMount.bind(null, testRoot, videos, simpleMapper), 'Able to render list of videos')
  t.true(testRoot.children[0].children[0].src.includes('youtube.com/embed/b8HO6hba9ZE'), 'videos are in page')
  t.end()
})

test('mixed mapper', function (t) {
  var testRoot = createTestElement()
  var mixedMapper = new Nanomap({
    video: YoutubeComponent,
    tweet: TwitterComponent,
    default: TwitterComponent
  })
  var videos = [
    'https://www.youtube.com/watch?v=b8HO6hba9ZE',
    'https://vimeo.com/229754542',
    'https://www.youtube.com/watch?v=bYpKrA233vY'
  ].map(function (url) {
    return { id: url, arguments: url, type: 'video' }
  })

  var tweets = [
    'https://twitter.com/uhhyeahbret/status/897603426518876161',
    'https://twitter.com/yoshuawuyts/status/895338700531535878'
  ].map(function (url) {
    return { id: url, arguments: url, type: 'tweet' }
  })

  delete tweets[0].type

  var data = videos.concat(tweets)

  t.doesNotThrow(renderAndMount.bind(null, testRoot, data, mixedMapper), 'Able to render list of videos')
  t.true(testRoot.children[0].children[0].src.includes('youtube.com/embed/b8HO6hba9ZE'), 'videos are in page')
  t.end()
})


example/simple/tweet-list.js
var Tweet = require('twitter-component')
var Nanocomponent = require('nanocomponent')
var assert = require('assert')
var html = require('bel')
var compare = require('nanocomponent/compare')
var Nanomap = require('../../')
var isArray = Array.isArray

function shapeTweetList (tweetUrl, i, list) {
  return {
    id: tweetUrl,
    opts: {
      placeholder: false
    },
    arguments: tweetUrl
  }
}

class TweetList extends Nanocomponent {
  constructor () {
    super()

    this.tweetList = null
    this.mapper = new Nanomap({ gc: false }, Tweet)
  }

  createElement (tweetList) {
    assert(isArray(tweetList), 'tweetList must be an array of tweet URLs')
    this.tweetList = tweetList.slice() // Have to slice since tweetList is an array
    return html`
      <div>
        ${tweetList.map(shapeTweetList).map(this.mapper)}
      </div>
    `
  }
  update (tweetList) {
    assert(isArray(tweetList), 'tweetList must be an array of tweet URLs')
    return compare(this.tweetList, tweetList)
  }
}

module.exports = TweetList


example/complex/mixed-list.js
var Tweet = require('twitter-component')
var YoutubeComponent = require('youtube-component')
var Nanocomponent = require('nanocomponent')
var assert = require('assert')
var html = require('bel')
var compare = require('nanocomponent/compare')
var Nanomap = require('../../')
var isArray = Array.isArray

function shapeData (url, i, list) {
  switch (true) {
    case /^https?:\/\/(www\.)?youtu\.be/i.test(url):
    case /^https?:\/\/(www\.)?youtube\.com/i.test(url):
    case /^https?:\/\/(www\.)?vimeo\.com/i.test(url):
    case /^https?:\/\/(www\.)?dailymotion\.com/i.test(url): {
      return {
        id: url,
        arguments: url,
        type: 'youtube-component',
        opts: {
          attr: {
            width: 480,
            height: 270
          }
        }
      }
    }
    case /^https?:\/\/(www\.)?twitter.com\/.*\/status\/\d*$/i.test(url): {
      return {
        id: url,
        arguments: url,
        type: 'twitter-component'
      }
    }
  }
}

class MixedList extends Nanocomponent {
  constructor () {
    super()

    this.urls = null
    this.mapper = new Nanomap({
      'youtube-component': YoutubeComponent,
      'twitter-component': Tweet
    })
  }

  createElement (urls) {
    assert(isArray(urls), 'MixedList: urls must be an array of tweet URLs')
    this.urls = urls.slice() // Have to slice since urls is an array
    return html`
      <div>
        ${urls.map(shapeData).map(this.mapper)}
      </div>
    `
  }
  update (urls) {
    assert(isArray(urls), 'tweetList must be an array of tweet URLs')
    return compare(this.urls, urls)
  }
}

module.exports = MixedList


example/complex/index.js
var html = require('choo/html')
var choo = require('choo')
var MixedList = require('./mixed-list')
var shuffleArray = require('fy-shuffle')

var app = choo()
app.use(tweetStore)
app.route('/', mainView)
if (typeof window !== 'undefined') {
  var container = document.createElement('div')
  container.id = 'container'
  document.body.appendChild(container)
  app.mount('#container')
}

var list = new MixedList()

function mainView (state, emit) {
  return html`
    <div id="container">
      <div>
        <button onclick=${ev => emit('shuffle-urls')}>shuffle urls</button>
        <button onclick=${ev => emit('reverse-urls')}>reverse urls</button>
        <button onclick=${ev => emit('add-url')}>add url</button>
        <button onclick=${ev => emit('append-url')}>append url</button>
        <button onclick=${ev => emit('pop-url')}>pop url</button>
        <button onclick=${ev => emit('shift-url')}>shift url</button>
        <button onclick=${ev => emit('re-render')}>plain render</button>
      </div>
      <div id="app">
        <h1>Embed some stuff with Choo</h1>
        ${list.render(state.tweets)}
      </div>
    </div>`
}

var moreTweets = [
  'https://twitter.com/uhhyeahbret/status/898315707254841344',
  'https://www.youtube.com/watch?v=b8HO6hba9ZE',
  'https://twitter.com/uhhyeahbret/status/898214560267608064',
  'https://vimeo.com/229754542',
  'https://twitter.com/uhhyeahbret/status/898196092189253632',
  'https://www.youtube.com/watch?v=bYpKrA233vY'
]

function tweetStore (state, emitter) {
  state.tweets = [
    'https://www.youtube.com/watch?v=wGCoAFZiYMw',
    'https://twitter.com/uhhyeahbret/status/897603426518876161',
    'https://twitter.com/yoshuawuyts/status/895338700531535878'
  ]

  emitter.on('DOMContentLoaded', function () {
    emitter.on('shuffle-urls', function () {
      state.tweets = shuffleArray(state.tweets)
      emitter.emit('render')
    })
    emitter.on('reverse-urls', function () {
      state.tweets = state.tweets.reverse()
      emitter.emit('render')
    })
    emitter.on('add-url', function () {
      var a = moreTweets.pop()
      if (a) {
        state.tweets.unshift(a)
        emitter.emit('render')
      }
    })
    emitter.on('append-url', function () {
      var a = moreTweets.pop()
      if (a) {
        state.tweets.push(a)
        emitter.emit('render')
      }
    })
    emitter.on('pop-url', function () {
      var a = state.tweets.pop()
      if (a) {
        moreTweets.push(a)
        emitter.emit('render')
      }
    })
    emitter.on('shift-url', function () {
      var a = state.tweets.shift()
      console.log(a)
      if (a) {
        moreTweets.push(a)
        emitter.emit('render')
      }
    })
    emitter.on('re-render', function () {
      emitter.emit('render')
    })
  })
}
